### ISSUE TITLE
Handle longtask attribution for embed & object in addition to iframe

### ISSUE THREAD

**Issue by spanicker** [WICG/longtasks#13]
**tl;dr**: In the current implementation TaskAttribution is populated with frameId, frameName, frameSrc etc pointing to the culprit iframe -- if culprit document frame is an iframe.  However we should also handle the case where the culprit is an embed or object.

**Details:**
"attribution" field is a set of "TaskAttributionType" entries - these are intended to contain things related to type of work eg. scriptUrl / src code location for script etc.
This [section in the explainer]( https://github.com/wicg/longtasks#pointing-to-the-culprit) provides relevant context here for the purpose & intention of "TaskAttributionType".
We added some fields in here pointing to the "culprit frame" (orthogonal to type of work) in which the script / layout executed. If the culprit is an iframe then "frameId", "frameName" & "frameSrc" are populated.

**Option #1:**
keep current flat structure in TaskAttributionTiming and add additional attributes such as:
- "containerType": depending on the culprit this would be set to: "frame", "object" or "embed" 
- If the above is "frame" then populate frameId, frameName, frameSrc as before; else if "object" then populate different fields: embedType, embedName, embedData etc; else if "embed" then ...

**Option #2:**
we could add a layer of abstraction and move everything related to the "container" to a "ContainerType" class, that way fields related to the "container" are separated from fields of "TaskAttributionType" -- which is intended to contain things related to type of work eg. scriptUrl etc.

Personally I am leaning towards #2.

@domenic
@igrigorik

**Comment by spanicker** [comment]
Updated V1 attribution doc with Option #2 proposal:
https://docs.google.com/document/d/1Yn_XA76DvZALiy24KkAqZVvLZuxqwD9X1F1oLftf6y4/edit#

**Comment by domenic** [comment]
In general I prefer flat structures to creating new classes. A class that just contains another class seems like a smell. Similarly, a "union type" where only some of its fields are applicable at a given time also seems not great. I would go with one of the following:

**Option (3):**

Recognize that embed and object are conceptually frames, and just use the existing attribute names, filling in appropriate values for the embed and object cases. So use frameSrc for embed's src attribute, etc. Optionally add a "containerElementName" or "containerType" field.

**Option (4):**

Find slightly more generic names and use them for everything. E.g. containerName, containerSrc, containerId, or sourceName, sourceSrc, sourceId. Again, optionally add a containerElementName or containerType field.

**Comment by igrigorik** [comment]
> Find slightly more generic names and use them for everything. E.g. containerName, containerSrc, containerId, or sourceName, sourceSrc, sourceId. Again, optionally add a containerElementName or containerType field.

👍

**Comment by spanicker** [comment]
Option #4 SGTM too.
We'd need something like containerElementName or containerType to indicate iframe vs. embed vs object.

**Comment by domenic** [comment]
I also realized that you can probably have `frame` elements trigger this. I haven't seen a `<frameset>` + `<frame>`s page in forever, but it could happen...

That shouldn't really change the discussion much, but I guess it's another case in the switch statement in the spec and implementation.

**Comment by spanicker** [comment]
Update spec & explainer with container* attributes

**Comment by spanicker** [comment]
Spec and implementation have been updated. Thanks all!
